ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,496 @@
+    lightweight.
+* Must be both thread-blocking/thread-safe and async-capable
+  * PoC shows that this is most easily done with Rust block/callback to `Queue` `push`/`pop` since in modern Ruby
+    `Queue.pop` does not block fibers and many other things don't either. Therefore many SDK features like clients and
+    codecs don't have to distinguish between sync/async.
+  * Activities need to work in both ways and _do_ have to make a distinction (discussed later).
+  * 💭 Why async-capable? Core and all SDKs are very parallel and so we want to give that benefit to users if they want
+    it.
+  * 💭 Why thread-blocking? That's the most common way Ruby is used.
+* Be as type-safe as possible with RBS
+  * All SDKs leverage generics to try to catch mistyped arguments for, say, calling a workflow/activity and we should
+    too.
+  * It is acknowledged that RBS and Ruby type hinting in general is rough at this time, but we must do our best.
+  * We may be able to leverage https://github.com/AaronC81/sord or similar to not have to duplicate docs and sig files,
+    but in general Ruby typing right now is a mess and we likely will hit bugs or advanced cases a yard/RBS combination
+    library won't work with.
+    * ❓ How hard do we want to work to avoid duplicate yard/RBS typing?
+  * 💭 Why not Sorbet? Sorbet can come later or provided separately, but RBS seems to be the newer movement.
+  * 💭 Why have any typing? As a library we might as well help users where we can.
+* Everything is in the `Temporalio` module and released as a `temporalio` gem starting at 0.2.0
+  * 💭 Why 0.2.0? 0.1.x versions already in use at https://rubygems.org/gems/temporalio.
+* Internal code will be in `Temporalio::Internal` module including bridge which is in `Temporalio::Internal::Bridge`
+  * 💭 Why? It is easier in yard to mark `@api private` on entire module and it avoids cluttering the API docs and other
+    modules with internal things.
+* Avoid cluttering the top-level `Temporalio` module with many small classes, but some are ok as we try to avoid a
+  `Temporalio::Common` module
+  * So `Temporalio::RetryPolicy` is preferred over `Temporalio::Common::RetryPolicy`.
+  * The search attribute collection is `Temporalio::SearchAttributes` and all things are under it, e.g.
+    `Temporalio::SearchAttributes::Key`.
+    * 💭 Why not `Temporalio::SearchAttribute` module with the collection underneath? The object has value on its own
+      so might as well make it a class.
+  * The base metric is `Temporalio::Metric`, and all metric things are under there, e.g. `Temporalio::Metric::Meter`.
+    * 💭 Why not `Temporalio::Metrics` module? The object has value on its own so might as well make it a class.
+   (e.g. `Temporalio::SearchAttribute::Key`)
+    and all metric things under `Temporalio::Metric` (e.g. `Temporalio::Metric::Meter`).
+  * 💭 Why no `Temporalio::Common`? Ruby expects to have common things where they make sense instead of under a `Common`
+    module.
```

Comment:
We may have to walk this back a bit. It is unreasonable to have `Temporalio::WorkflowIDReusePolicy`, `Temporalio::WorkflowIDConflictPolicy`, etc cluttering the top-level. However, we don't want `RetryPolicy` to be in `Common`. So I guess we may need some kinda hybrid rule that says use the `Common` module for _small_ or _lesser used_ common classes?